File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/02.05.Animals/Animal.php
<?php

require_once "SoundProducible.php";
abstract class Animal implements SoundProducible
{
    const INVALID_INPUT_MSG = "Invalid input!";
    /**
     * @var string
     */
    private $name;
    /**
     * @var int
     */
    private $age;
    /**
     * @var string
     */
    private $gender;

    /**
     * Animal constructor.
     * @param string $name
     * @param int $age
     * @param string $gender
     * @throws Exception
     */
    public function __construct(string $name, int $age, string $gender)
    {
        $this->setName($name);
        $this->setAge($age);
        $this->setGender($gender);
    }

    /**
     * @param string $name
     * @throws Exception
     */
    public function setName(string $name): void
    {
        if (empty(trim($name))) {
            throw new Exception(self::INVALID_INPUT_MSG);
        }
        $this->name = $name;
    }

    /**
     * @param int $age
     * @throws Exception
     */
    public function setAge(int $age): void
    {
        if (empty(trim($age)) || $age <= 0) {
            throw new Exception(self::INVALID_INPUT_MSG);
        }
        $this->age = $age;
    }

    /**
     * @param string $gender
     * @throws Exception
     */
    public function setGender(string $gender): void
    {
        if (empty(trim($gender))) {
            throw new Exception(self::INVALID_INPUT_MSG);
        }
        $this->gender = $gender;
    }
    public function __toString()
    {
        $output =  " {$this->name} {$this->age} {$this->gender}" . PHP_EOL . $this->produceSound();;
        return $output;
    }
}

File: Courses/C# OOP/Basics/10. Inheritance Exercise/Exercises_Inheritance/06.Animals/Animals/Animal.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace _06.Animals.Animals
{
    public abstract class Animal
    {
        private string name;
        private int age;
        private string gender;

        public Animal(string name, int age, string gender)
        {
            Name = name;
            Age = age;
            Gender = gender;
        }

        public string Name
        {
            get => name; 
            private set
            {
                if (string.IsNullOrEmpty(value) || string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException("Invalid input!");
                }
                name = value;
            }
        }
        public int Age
        {
            get => age;
            private set
            {
                if (value < 0)
                {
                    throw new ArgumentException("Invalid input!");
                }
                age = value;
            }
        }
        public string Gender
        {
            get => gender; 
            private set
            {
                if (string.IsNullOrEmpty(value) || string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException("Invalid input!");
                }
                gender = value;
            }
        }

        public virtual void ProduceSound()
        {
            Console.WriteLine("Main function");
        }
        public override string ToString()
        {
            return $"{Name} {Age} {Gender}";
        }
    }
}
